### ISSUE TITLE
ForwardDiff + destructure is different from Zygote, on a model with BatchNorm

### ISSUE THREAD

**Issue by lazarusA** [FluxML/Flux.jl#2122]
### Package Version

Optimisers v0.2.10, ForwardDiff v0.10.33, Flux v0.13.7

### Julia Version

1.8

### OS / Environment

OS

### Describe the bug

The example from Optimisers using ForwardDiff compared with the output from Zygote is unfortunately not the same.

### Steps to Reproduce

```Julia
using ForwardDiff  # an example of a package which only likes one array
using Flux
using Random
using Optimisers
Random.seed!(123)

model = Chain(  # much smaller model example, as ForwardDiff is a slow algorithm here
          Conv((3, 3), 3 => 5, pad=1, bias=false), 
          BatchNorm(5, relu), 
          Conv((3, 3), 5 => 3, stride=16),
        )
image = rand(Float32, 224, 224, 3, 1);
@show sum(model(image));

loss(m, x) = sum(m(x))

rule = Optimisers.Adam(0.001f0,  (0.9f0, 0.999f0), 1.1920929f-7)

flat, re = Flux.destructure(model)
st = Optimisers.setup(rule, flat)  # state is just one Leaf now

∇flat = ForwardDiff.gradient(flat) do v
    loss(re(v), image) # re(v), rebuild a new object like model
end

st, flat = Optimisers.update(st, flat, ∇flat)
@show loss(re(flat),image);
```

and Zygote version

```Julia
using Flux
using Random
Random.seed!(123)

model = Chain(  # much smaller model example, as ForwardDiff is a slow algorithm here
          Conv((3, 3), 3 => 5, pad=1, bias=false), 
          BatchNorm(5, relu), 
          Conv((3, 3), 5 => 3, stride=16),
        )
image = rand(Float32, 224, 224, 3, 1);
@show sum(model(image));

loss(m, x) = sum(m(x))

opt = Flux.Adam(0.001f0,  (0.9f0, 0.999f0), 1.1920929f-7)
θ = Flux.params(model)
grads = Flux.gradient(θ) do 
    loss(model, image)
end

Flux.update!(opt, θ, grads)
@show loss(model, image);
```

### Expected Results

```
sum(model(image)) = -0.33076355f0
loss(model, image) = -5.064876f0
```

### Observed Results

```
sum(model(image)) = -0.33076355f0
loss(re(flat), image) = -7.7023053f0
```

**Comment by ToucheSir** [comment]
AbstractDiff is pretty similar. Let me file an issue over there and see how it goes. We can always look into the NNlib option in parallel.

**Comment by mcabbott** [comment]
Besides detecting whether you are within AD (also an issue for dropout), the problem with BatchNorm is that ForwardDiff runs the forward pass several times (chunked mode, for any large array).

I can't think of a good way to detect that. Perhaps we should make it a clear error instead?

**Comment by CarloLucibello** [comment]
Revisiting on Flux 0.16 / ForwardDiff 1.4 / NNlib 0.9.

The **ForwardDiff gradient itself is correct** — numerically identical to Zygote's (relative error ~3e-7, i.e. `Float32` roundoff). The difference in the reproducer is not in the gradient.

It comes entirely from BatchNorm's **running-statistics update**, which is a mutating side-effect of the forward pass:

- Zygote runs the forward once → running stats get **one** EMA update.
- ForwardDiff runs the forward `⌈N/chunksize⌉` times (chunked mode) → running stats get updated **that many times**.

The update count tracks the number of forward passes exactly (here `N = 283` params):

| AD / chunk size | forward passes | BN `σ²` after |
|---|---|---|
| Zygote | 1 | 0.907 |
| ForwardDiff, `Chunk{280}` | 2 | 0.824 |
| ForwardDiff, `Chunk{32}` | 9 | 0.431 |
| ForwardDiff, `Chunk{10}` | 29 | 0.115 |
| ForwardDiff, `Chunk{1}` | 283 | 0.071 |

The reproducer's final `loss(re(flat), image)` runs in **test mode**, so it uses those over-updated running stats — hence the different number. In *training* mode the forward output depends only on the batch statistics (not the running stats), so every chunk produces the same output and the gradient is unaffected.

This is an inherent ForwardDiff limitation: it evaluates the (side-effecting) forward pass multiple times, and there is no seam to make the running-stats mutation happen exactly once. Skipping the update under ForwardDiff wouldn't reproduce Zygote's once-per-step semantics either.

**Workarounds:**
- Call `testmode!(model)` before differentiating if you rely on the running stats — there is then no tracking, and the ForwardDiff gradient is clean, deterministic, and identical to Zygote (verified: `σ²` untouched, gradient matches to ~3e-7).
- Otherwise the gradient is already correct; just don't rely on BatchNorm's running stats being updated exactly once under ForwardDiff.
- For training models with BatchNorm, prefer Zygote or Enzyme.

Closing as not planned, since this is inherent to ForwardDiff. Happy to reopen if someone wants a warning/error for `track_stats=true` + training mode under ForwardDiff.
